File: others/pitchfork-reviews.py
import gevent
from gevent import monkey; monkey.patch_all()
from gevent.pool import Pool
from pyquery import PyQuery as pq
import sys
from itertools import imap, ifilter
from pprint import pprint
import urllib2

def album(el):
    li_tag = pq(el)
    return {
        "artist": li_tag("h1").text(),
        "title": li_tag("h2").text(),
        "url": "http://pitchfork.com" + li_tag("a").attr("href"),
        "score": None,
    }

def downloadPQ(url):
    return pq(urllib2.urlopen(url).read())

def downlaadAlbums(page):
    return imap(
        album, 
        downloadPQ("http://pitchfork.com/reviews/albums/{page}".format(page=page))("ul.object-grid ul li")
    )

def downloadScore(url):
    try:
        return float(downloadPQ(url)("span.score").text())
    except:
        return None


def setScore(score, album):
    album['score'] = score
    return album


def display(albums):
    for album in albums:
        print "{score} {artist} - {title} {url}".format(**album)


def main():
    try:
        page = int(sys.argv[1])
    except:
        page = 1

    albums = list(downlaadAlbums(page))
    pool = Pool()
    scores = pool.map(lambda a: downloadScore(a['url']), 
                      albums)
    display(
        sorted(
            ifilter(
                lambda a: a['score'] > 7.0,
                imap(
                    setScore,
                    scores,
                    albums)),
            key=lambda a: a['score'],
            reverse=True
            ),
        )
                    
                    
if __name__ == '__main__': main()

File: others/pitchfork-reviews.js
var http = require("http");
var Q = require("q");
var $ = require("jquery");


function downloadJQ(url) {
    var deferred = Q.defer();
    http.get(
	url, function(res) {
	    var body = "";
	    res.on("data", function(chunk) {
		body += chunk;
	    });

	    res.on("end", function() {
		deferred.resolve($(body));
	    })
	}
    );
    return deferred.promise;
}

function album(li_tag) {
    var $li_tag = $(li_tag);
    return {
	"artist": $li_tag.find("h1").text(),
	"title": $li_tag.find("h2").text(),
	"url": "http://pitchfork.com" + $li_tag.find("a").attr("href"),
	"score": null
    };
}

function trace(val) {
    console.dir(val);
    return val;
}

function downloadAlbums(page) {
    return downloadJQ("http://pitchfork.com/reviews/albums/" + page + "/").then(
	function($d) {
	    return $d.find("ul.object-grid ul li").map(function(i, el) { return album(el); });
	});
}

function downloadScore(a) {
    return downloadJQ(a['url']).then(
	function($d) {
	    return setScore(
		parseFloat($d.find("span.score").text()),
		a);
	});
}

function setScore(score, album) {
    album['score'] = score
    return album
}

function cmp(x,y) {
    if(x == y) { 
	return 0;
    } else if (x < y) {
	return -1;
    } else {
	return 1;
    }
}

function display(albums) {
    albums.forEach(function(result) {
	var album = result.value;
	if(album['score'] > 7) {
	    console.log(
		    album['score']
		    + " "
		    + album['artist'] 
		    + " - " 
		    + album['title']
		    + album['url']
	    )
	}
    });
}

function main() {
    downloadAlbums(1).then(function(albums) {
	return Q.allSettled(
	    albums.map(function(i, a) { return downloadScore(a) })
	).then(function(albums) {
	    display(albums.sort(function(y,x) { cmp(x['value']['score'], y['value']['score']) }));
	});
    }).fail(console.error);
}

main();
